File: NyaLauncher.Core/Download/VersionFilter.cs
using NyaLauncher.Core.Models;

namespace NyaLauncher.Core.Download;

/// <summary>
/// Minecraft 版本筛选服务
/// </summary>
public static class VersionFilter
{
    /// <summary>
    /// 根据筛选类型过滤版本列表
    /// </summary>
    /// <param name="versions">原版本列表</param>
    /// <param name="filter">筛选键: "all" 全部, "release" 正式版, "snapshot" 快照版, "old" 远古版本</param>
    /// <returns>过滤后的版本列表</returns>
    public static List<MinecraftVersion> Apply(List<MinecraftVersion> versions, string filter)
    {
        return filter switch
        {
            "release"  => [.. versions.Where(v => v.Type == "release")],
            "snapshot" => [.. versions.Where(v => v.Type == "snapshot")],
            "old"      => [.. versions.Where(v => v.Type is "old_beta" or "old_alpha")],
            _          => versions
        };
    }
}

File: internal/download/version_filter.go
package download

import "nyalauncher/internal/models"

// VersionFilter Minecraft 版本筛选服务。对应 C# VersionFilter。

// ApplyVersionFilter 根据筛选类型过滤版本列表。
// filter: "all" 全部, "release" 正式版, "snapshot" 快照版, "old" 远古版本。
func ApplyVersionFilter(versions []models.MinecraftVersion, filter string) []models.MinecraftVersion {
	switch filter {
	case "release":
		return filterVersions(versions, func(v models.MinecraftVersion) bool { return v.Type == "release" })
	case "snapshot":
		return filterVersions(versions, func(v models.MinecraftVersion) bool { return v.Type == "snapshot" })
	case "old":
		return filterVersions(versions, func(v models.MinecraftVersion) bool {
			return v.Type == "old_beta" || v.Type == "old_alpha"
		})
	default:
		return versions
	}
}

func filterVersions(versions []models.MinecraftVersion, predicate func(models.MinecraftVersion) bool) []models.MinecraftVersion {
	result := make([]models.MinecraftVersion, 0, len(versions))
	for _, v := range versions {
		if predicate(v) {
			result = append(result, v)
		}
	}
	return result
}
